### PR TITLE
Support Asynchronous Writing Datastores

### PR DESCRIPTION
**Pull request by aschmahmann**
Per #137 we would like to support Datastores that internally have asynchronous writes. We are doing this by adding a `Sync(prefix ds.Key)` function to the Datastore interface.

The plan is:
1) Get this PR reviewed for a sanity check
2) Get the datastore PRs submitted and reviewed and ready to be accepted (their go.mod files will reference this PR's branch)
   - [x] https://github.com/ipfs/go-ds-measure/pull/23
   -  [x] https://github.com/ipfs/go-ds-s3/pull/46
   -  [x] https://github.com/ipfs/go-ds-redis/pull/6
   -  [x] https://github.com/ipfs/go-ds-leveldb/pull/36
   -  [x] https://github.com/ipfs/go-ds-flatfs/pull/60
   -  [x] https://github.com/ipfs/go-ds-badger/pull/77
   - Note: If some of the datastores are proving tricky to write a Sync command for we may not block for them all to be approved before we continue.
3) Merge this PR
4) Update and merge the other PRs
5) Update some areas of go-ipfs to use the `Sync()` command on their datastores
6) Switch the default datastore in go-ipfs to be an asynchronous one (i.e. Badger with Async writes enabled)

### CODE REVIEW

File: examples/fs.go
Comment (Stebalien) [inline examples/fs.go:112]: (this may break in weird cases (.., etc.). but this is just an example.

File: mount/mount.go
Comment (Stebalien) [inline mount/mount.go]: We need to sync _all_ datastores under the prefix. This will only sync the one responsible for the given key.

If we have two mounts `["/", "/blocks"]`, `Sync("/")` will currently _just_ sync "/".

Comment (Stebalien) [reply mount/mount.go]: Maybe we can modify the test datastore to record all `Sync` calls then use that o make sure sync is getting called correctly?

**Review by Stebalien — CHANGES_REQUESTED** [review]
The mount datastore needs a small fix and a test but otherwise LGTM.

File: mount/mount.go
Comment (aschmahmann) [reply mount/mount.go]: ah cool, I didn't realize we wanted mount to be able to do that. I figured that if we wanted submounts that the "/foo" would point to another mount datastore that was "/foo/bar".

should be fixable though. will take a look later today

File: examples/fs.go
Comment (aschmahmann) [reply examples/fs.go:112]: Yep, I think for the example datastore not worth thinking too hard about. If similar issues come up in FlatFS we'll deal with it there.

File: mount/mount.go
Comment (Stebalien) [reply mount/mount.go]: Yeah, `lookup` is just for finding the datastore where we should read/write a key.

File: mount/mount.go
Comment (Stebalien) [inline mount/mount.go]: I'd expect syncing a non-existent key to return success: there's nothing there so there's nothing to persist.

`Put` returns an error because the put fails. `Get` just returns `ErrNotFound`.

File: mount/mount.go
Comment (Stebalien) [inline mount/mount.go]: Can we use `lookupAll`?

**Review by Stebalien — CHANGES_REQUESTED** [review]

**Comment by bonedaddy** [comment]
After this PR is merged can we have a version bump according to semver practices since this is a breaking change to the `Datastore` interface?

**Comment by Stebalien** [comment]
> After this PR is merged can we have a version bump according to semver practices since this is a breaking change to the Datastore interface?

We will bump the minor version because the package version is sub-zero.

**Comment by bonedaddy** [comment]
> > After this PR is merged can we have a version bump according to semver practices since this is a breaking change to the Datastore interface?
> 
> We will bump the minor version because the package version is sub-zero.

Won't break `go get -u=patch` so :rocket:

File: mount/mount.go
Comment (aschmahmann) [reply mount/mount.go]: No problem here.

My thought was that if you tried to access an unmounted part of the hierarchy you might want to get an error. It's reasonable (and easier for testing purposes) to say that we did Sync all puts/deletes under that key and there happen to be none there because that key is unmounted.

File: mount/mount.go
Comment (aschmahmann) [reply mount/mount.go]: np, it might do a few more comparisons but it'll look much cleaner

Comment (aschmahmann) [reply mount/mount.go]: lmk how it looks now

File: mount/mount.go
Comment (Stebalien) [inline mount/mount.go:204]: Hm. This looks like a bug in `lookupAll`. `lookupAll` should only return mounts that can contain descendents of `key`.

**Review by Stebalien — APPROVED** [review]

File: mount/mount.go
Comment (Stebalien) [reply mount/mount.go:204]: We can merge this as-is and fix it later or move this logic into `lookupAll` now.

Comment (aschmahmann) [reply mount/mount.go:204]: Could you clarify what you think the bug is here? It looks like the documented behavior for `lookupAll`.

Comment (Stebalien) [reply mount/mount.go:204]: `lookupAll` is supposed to return all datastores that _might_ contain keys that are descendants of the given key. That means we should be calling `Sync` on all datastores returned by `lookupAll` datastores and we shouldn't have this extra check.

However, if we have `["/foo", "/foo/bar"]`, `"/foo"` should "cover" `"/foo/bar"` given that `lookup` will always return `"/foo"`. However, when given the prefix `"/foo"`, `lookupAll` will return _both_.

Note: this is only _technically_ a bug. Nobody would ever mount `"/foo"` over `"/foo/bar"`.

Comment (aschmahmann) [reply mount/mount.go:204]: > Nobody would ever mount "/foo" over "/foo/bar"

I thought it was expected behavior that we might mount "/" and also "/foo", or is the root a special case?

Comment (Stebalien) [reply mount/mount.go:204]: My first example was bad: order matters. If we mount "/" over "/foo", "/" will win (I think? check me here). In go-ipfs, we mount "/blocks" over "/".

Comment (Stebalien) [reply mount/mount.go:204]: I say "nobody would ever mount "/foo" over "/foo/bar" because "/foo" is a prefix of "/foo/bar" (and we check mounts in-order).

**Comment by Stebalien** [comment]
LGTM but let's merge this tomorrow. Hitting merge then going to bed is a bad idea (I mean, it's one of my favorite bad ideas but I think it's time to call it a day).

File: mount/mount.go
Comment (Stebalien) [reply mount/mount.go:204]: I was wrong. We reverse-sort datastores so "/foo/bar" always comes before "/foo".

However, the bug was valid. Fix in #141.

**Pull request merged**
